File: examples/experiments/functions/src/jsonparse.go
package main

// jsonparse — SINTETICA. Serializzazione e deserializzazione JSON ripetuta.
//
// Profilo atteso: CPU utente alta accompagnata da page fault elevati. La
// deserializzazione alloca molti oggetti di piccola dimensione, il garbage
// collector interviene di frequente, e la memoria viene continuamente
// richiesta e restituita al sistema operativo.
//
// E' un profilo che nessuna funzione esistente occupa: randomaccess ha page
// fault alti ma su un'area preallocata, senza pressione sull'allocatore.

import (
	"encoding/json"
	"fmt"
	"runtime"

	"github.com/serverledge-faas/serverledge/serverledge"
)

type record struct {
	ID       int               `json:"id"`
	Name     string            `json:"name"`
	Tags     []string          `json:"tags"`
	Metadata map[string]string `json:"metadata"`
	Values   []float64         `json:"values"`
}

func myHandler(params map[string]interface{}) (interface{}, error) {
	records := 20000
	if val, ok := params["records"].(float64); ok {
		records = int(val)
	}

	// Ridotto da 20 a 8 in fase di taratura.
	rounds := 8
	if val, ok := params["rounds"].(float64); ok {
		rounds = int(val)
	}

	source := make([]record, records)

	for i := range source {
		source[i] = record{
			ID:   i,
			Name: fmt.Sprintf("entity-%d", i),
			Tags: []string{"alpha", "beta", "gamma"},
			Metadata: map[string]string{
				"region": "eu",
				"tier":   fmt.Sprintf("t%d", i%5),
			},
			Values: []float64{float64(i), float64(i) * 1.5, float64(i) * 2.25},
		}
	}

	totalBytes := 0
	totalParsed := 0

	for round := 0; round < rounds; round++ {
		encoded, err := json.Marshal(source)
		if err != nil {
			return nil, err
		}

		totalBytes += len(encoded)

		// Il risultato viene deserializzato in una struttura nuova a ogni
		// giro: riusare la stessa slice ridurrebbe le allocazioni e
		// snaturerebbe il profilo che si vuole ottenere.
		var decoded []record
		if err := json.Unmarshal(encoded, &decoded); err != nil {
			return nil, err
		}

		totalParsed += len(decoded)
	}

	return map[string]interface{}{
		"message":      "JSON round-trip completed",
		"records":      records,
		"rounds":       rounds,
		"total_parsed": totalParsed,
		"total_mb":     float64(totalBytes) / 1024 / 1024,
		"arch":         runtime.GOARCH,
	}, nil
}

func main() {
	serverledge.Start(myHandler)
}

File: examples/experiments/multilang/python/jsonparse.py
import json


def _int_param(params, name, default, minimum, maximum):
    try:
        value = int(params.get(name, default))
    except (TypeError, ValueError, AttributeError):
        value = default
    return max(minimum, min(maximum, value))


def handler(params, context):
    params = params or {}

    records = _int_param(params, "records", 20_000, 100, 100_000)
    rounds = _int_param(params, "rounds", 20, 1, 100)

    source = []

    for i in range(records):
        source.append(
            {
                "id": i,
                "name": f"entity-{i}",
                "tags": ["alpha", "beta", "gamma"],
                "metadata": {"region": "eu", "tier": f"t{i % 5}"},
                "values": [float(i), float(i) * 1.5, float(i) * 2.25],
            }
        )

    total_bytes = 0
    total_parsed = 0
    checksum = 0

    for _ in range(rounds):
        encoded = json.dumps(source, separators=(",", ":"), ensure_ascii=False).encode("utf-8")
        total_bytes += len(encoded)

        decoded = json.loads(encoded)
        total_parsed += len(decoded)

        if decoded:
            checksum = (
                checksum + decoded[0]["id"] + decoded[-1]["id"] + len(decoded[-1]["tags"])
            ) & 0xFFFFFFFF

    return {
        "benchmark": "jsonparse",
        "records": records,
        "rounds": rounds,
        "total_parsed": total_parsed,
        "total_mb": total_bytes / 1024.0 / 1024.0,
        "checksum": checksum,
    }
